File: flask/blackjack.py
from flask import current_app as app
from flask import render_template, request, flash, redirect, url_for, json
from models import UserInfo
from main import db, socketio
from flask_login import login_user, logout_user, login_required, current_user
from flask_socketio import SocketIO, emit, disconnect
import random

cards = ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K']
cards_player = {}
cards_dealer = {}
bet = {}

def blackjack_sum(cards_list):
    soft_sum = 0
    num_of_A = 0
    for i in cards_list:
        if i == 'A':
            soft_sum += 11
            num_of_A += 1
        elif i == 'J' or i == 'Q' or i == 'K':
            soft_sum += 10
        else:
            soft_sum += int(i)
    
    while soft_sum > 21 and num_of_A > 0:
        num_of_A -= 1
        soft_sum -= 10
    return soft_sum


@app.route('/blackjack_rules', methods=['GET'])
@login_required
def blackjack_rules():
  '''
  Displays the game's rules (accessible from main menu).
  '''
  user = current_user
  return render_template('blackjack_rules.html', username=user.username, coins=user.coins)

@app.route('/blackjack_play', methods=['GET'])
@login_required
def blackjack_play():
    user = current_user
    #return render_template('blackjack_play.html', username=user.username, coins=user.coins)
    return render_template('blackjack_play.html', async_mode=socketio.async_mode, username=user.username, coins=user.coins)

@socketio.on('connect_event', namespace='/blackjack')
def test_message(message):
    emit('server_response', {'data': message['data']})

@socketio.on('client_event', namespace='/blackjack')
def test_client_message(message):
    emit('server_response', {'data': message['data']+' Received.'})

@socketio.on('disconnect_request', namespace='/blackjack')
def test_disconnect_request():    
    def can_disconnect():
        disconnect()
    emit('server_response', {'data': 'Disconnected!'}, callback=can_disconnect)

@socketio.on('connect', namespace='/blackjack')
def test_connect():
    emit('server_response', {'data': 'Server Connected!'})

@socketio.on('disconnect', namespace='/blackjack')
def test_disconnect():
    print('Client disconnected', request.sid)

@socketio.on('start_game', namespace='/blackjack')
def start_game(message):
    global bet
    global cards_player
    global cards_dealer
    user = current_user
    bet[user.username] = message['bet_coin']
    cards_player[user.username] = []
    cards_dealer[user.username] = []
    cards_player[user.username] = random.choices(cards, k=2)[:]
    cards_dealer[user.username] = random.choices(cards, k=2)[:]
    cards_dealer_show = cards_dealer[user.username].copy()
    cards_dealer_show[0] = 'X'
    if blackjack_sum(cards_player[user.username]) == 21 and blackjack_sum(cards_dealer[user.username]) == 21:
        emit('start', {'player': cards_player[user.username], 'dealer': cards_dealer[user.username]})
        emit('server_response', {'data': 'Draw!'})
        emit('draw')
        emit('finish_game', {'res': 'Draw'})
    elif blackjack_sum(cards_player[user.username]) == 21:
        emit('start', {'player': cards_player[user.username], 'dealer': cards_dealer[user.username]})
        user.coins += round(int(bet[user.username])*1.5)
        emit('server_response', {'data': 'You Win! BlackJack!'})
        emit('blackjack')
        emit('finish_game', {'res': 'Win ' + str(int(bet[user.username])*1.5) + ' Coins'})
    elif blackjack_sum(cards_dealer[user.username]) == 21:
        emit('start', {'player': cards_player[user.username], 'dealer': cards_dealer[user.username]})
        user.coins -= int(bet[user.username])   
        emit('server_response', {'data': 'Dealer Win!'})
        emit('dealer_win')         
        emit('finish_game', {'res': 'Lose ' + bet[user.username] + ' Coins'})
    else:
        emit('start', {'player': cards_player[user.username], 'dealer': cards_dealer_show})
    
    db.session.commit()

@socketio.on('player_take_one', namespace='/blackjack')
def player_take_one():
    user = current_user
    card = random.choice(cards)
    cards_player[user.username].append(card)
    emit('player_picked', {'cards': cards_player[user.username]})
    if blackjack_sum(cards_player[user.username]) > 21:
        emit('player_bust')

@socketio.on('player_stop', namespace='/blackjack')
def player_stop():
    global cards_dealer
    user = current_user

    while blackjack_sum(cards_dealer[user.username]) < 17:
        cards_dealer[user.username].append(random.choice(cards))
    emit('dealer_picked', {'cards': cards_dealer[user.username], 'total': blackjack_sum(cards_dealer[user.username])})

    player_total = blackjack_sum(cards_player[user.username])
    dealer_total = blackjack_sum(cards_dealer[user.username])
    if player_total > 21:
        user.coins -= int(bet[user.username])
        emit('server_response', {'data': 'Dealer Win!'})
        emit('dealer_win')
        emit('finish_game', {'res': 'Lose ' + bet[user.username] + ' Coins'})
    elif dealer_total > 21:
        user.coins += int(bet[user.username])
        emit('server_response', {'data': 'You Win!'})
        emit('player_win')
        emit('finish_game', {'res': 'Win ' + bet[user.username] + ' Coins'})
    else:
        if dealer_total > player_total:
            user.coins -= int(bet[user.username])   
            emit('server_response', {'data': 'Dealer Win!'})
            emit('dealer_win')         
            emit('finish_game', {'res': 'Lose ' + bet[user.username] + ' Coins'})
        elif player_total > dealer_total:
            user.coins += int(bet[user.username])     
            emit('server_response', {'data': 'You Win!'})
            emit('player_win')
            emit('finish_game', {'res': 'Win ' + bet[user.username] + ' Coins'})
        else:
            emit('server_response', {'data': 'Draw!'})
            emit('draw')
            emit('finish_game', {'res': 'Draw'})

    db.session.commit()

File: flask/static/js/blackjack.js
function add_card(who, card){
    var new_card = $("#example-card").clone();
    new_card.addClass("added");
    if (card!="X")
        new_card.attr("src","static/images/cards/"+card+"C"+".png");
    else
        new_card.addClass("hidden");
    if (who=="dealer")
        new_card.appendTo( "#dealer" );
    if (who=="player")
        new_card.appendTo( "#player" );
    new_card.show();       
}
$(document).ready(function(){
    namespace = '/blackjack';
    var socket = io(namespace);
    var coins_now = Number($("#coins_store").val());
    var username = $("#username_store").val();
    var bet_show = 0;

    socket.on('connect', function(){
        socket.emit('connect_event', {'data': username + ' Connected!'});
    });

    socket.on('start', function(msg){
        $( ".added" ).remove();
        add_card("player",msg.player[0]);
        add_card("dealer",msg.dealer[0]);
        add_card("player",msg.player[1]);
        add_card("dealer",msg.dealer[1]);
        // document.getElementById("dealer").innerHTML=msg.dealer;
        $('form#take').find(':input[type=submit]').prop('disabled', false);
        $('form#stop').find(':input[type=submit]').prop('disabled', false);
        // document.getElementById("log").innerHTML='';
        // $('#log').append('<br>' + $('<div/>').text('Received #: Start a new game.').html());
    });

    socket.on('server_response', function(msg, cb){
        // $('#log').append('<br>' + $('<div/>').text('Received #' + ': ' + msg.data).html());
        // if(cb)
        //     cb();
    });

    socket.on('player_picked', function(msg){
        // $('#log').append('<br>' + $('<div/>').text('Received #: Player drew a card.').html());
        // document.getElementById("player").innerHTML=msg.cards;
        add_card("player",msg.cards[msg.cards.length-1]);
    });

    socket.on('player_bust', function(){
        // $('#log').append('<br>' + $('<div/>').text('Player Bust!').html());
        socket.emit('player_stop');
    });

    socket.on('player_21', function(){
        // $('#log').append('<br>' + $('<div/>').text('Player got 21!').html());
        alert("player 21!");
        socket.emit('player_stop');                
    });

    socket.on('dealer_picked', function(msg){
        $(".hidden").attr("src","static/images/cards/"+msg.cards[0]+"C"+".png");
        for ( var i = 2; i < msg.cards.length; i++ ) {
            add_card("dealer",msg.cards[i]);
        }
        
    });

    socket.on('finish_game', function(msg){
        bet_show = 0;
        // document.getElementById("show_result").innerHTML= "<h4>Last time: " + msg.res + "</h4>";
        $('form#take').find(':input[type=submit]').prop('disabled', true);
        $('form#stop').find(':input[type=submit]').prop('disabled', true);
        $('form#start').find(':input[type=submit]').prop('disabled', false);
        $('form#start').find(':input[type=number]').prop('disabled', false);
        $('#user_coins').html("Coins: "+coins_now.toString());
        // document.getElementById("head").innerHTML='<h1>CSIE Online Games --- Black Jack</h1><h4>User: '+username+'<br> Coins: '+coins_now+'</h4>'
    });

    socket.on('draw', function(){
        alert('Draw!');
    });

    socket.on('player_win', function(){
        alert('You Win!')
        coins_now += Number(bet_show)
    });

    socket.on('dealer_win', function(){
        alert('Dealer Win!')
        coins_now -= Number(bet_show)
    });

    socket.on('blackjack', function(){
        alert('You Win!<br>Black Jack!')
        coins_now += Math.round(Number(bet_show)*1.5)
    });

/*
    $('form#emit').submit(function(event){
        socket.emit('client_event', {'data': $('#emit_data').val()});
        return false;
    });

    $('form#disconnect').submit(function(event){
        socket.emit('disconnect_request');
        return false;
    });
*/
    $('form#start').submit(function(event){
        $('form#start').find(':input[type=submit]').prop('disabled', true);
        $('form#start').find(':input[type=number]').prop('disabled', true);
        bet_show = $('#bet').val();
        socket.emit('start_game', {'bet_coin': bet_show});
        return false;
    });

    $('form#take').submit(function(event){
        socket.emit('player_take_one');
        return false;
    });

    $('form#stop').submit(function(event){
        socket.emit('player_stop');
        return false;
    });

});